### DISCUSSION TITLE
Issue with the efficiency of Distributed mesh with mesh adaptivity

### DISCUSSION THREAD

**Discussion by simopier** [idaholab/moose#18163] [Q&A Meshing] [answered]
When performing large 3D phase field simulations, the computational cost can quickly become prohibitive. For that reason, I have been trying to leverage some of the features offered by MOOSE to decrease these costs. Two of them being the use of Distributed mesh over Replicated mesh, and the use of mesh adaptivity. 

However, I do not get good performances when I try to combined the two, especially when my phase field simulation uses elasticity. After discussing this with @roystgnr, @amjokisaari, and @jiangwen84, I figured I would provide examples of the types of system I am trying to simulate, and document the performance issues I was observing.

I am performing phase field simulation with elasticity, and below are the different combinations of options that I have used: 
1. Using Distributed mesh or Replicated mesh
2. Using mesh adaptivity (2 levels) or not.
3. When using Distributed mesh, I also tried using the `part_package = ptscotch` option.

I have recorded the active time for each of these 6 simulations using the postprocessor
```
[./activetime]  
    type = PerfGraphData  
    data_type = TOTAL  
    section_name = Root  
[../]
```

The results are provided in the table below. I made sure to perform large enough simulations to have a significant amount of nonlinear DOFs (> 2M) for Distributed mesh to be relevant. Note that when using mesh adaptivity (MA), I simulated a larger domain to still have a large amount of DOFs. Results show that without mesh adaptivity (left figure), the active time does decrease when using the distributed mesh rather than the regular mesh. However, the trend is reversed when using mesh adaptivity (right figure). With mesh adaptivity, using Distributed mesh is more time consuming that using Replicated mesh. I am not sure what causes this, but it limits the advantage of Distributed mesh. 
 
| Without mesh adaptivity      | With mesh adaptivity (larger domain)     |
|-------------|-------------|
| <img width="549" alt="0_El" src="https://user-images.githubusercontent.com/17445651/123294780-62c17b80-d4d2-11eb-9524-e45b4d58ae03.png"> | <img width="550" alt="0_El_MA" src="https://user-images.githubusercontent.com/17445651/123294807-69e88980-d4d2-11eb-98fc-955e7d2fb35b.png"> |

Attached with this post are the input files, and example of the .pbs file that I have been using to run these on HPC (falcon), and the output files. Note that I have run these simulations with the `-snes_view -log_view` options, as well as with 
```
[./pgraph]
  type = PerfGraphOutput
  execute_on = 'initial final'  # Default is "final"
  level = 3                     # Default is 1
  heaviest_branch = true        # Default is false
  heaviest_sections = 7         # Default is 0
 []
```
to provide performance data. Let me know I need to provide more information or run additional tests.
[Moose_discussion_simopier_distributed_mesh.zip](https://github.com/idaholab/moose/files/6710578/Moose_discussion_simopier_distributed_mesh.zip)

**Comment by lynnmunday** [comment]
Why does distributed run faster than replicated?  I thought the only advantage distributed had over replicated was a reduction in memory.

**Reply by roystgnr** [reply]
That's a very good question.  DistributedMesh setup can theoretically be faster, since you're not setting up a bunch of remote elements on each processor that you'll never use, but after that point the only advantage in CPU speed ought to be when the mesh is modified, which shouldn't affect the non-adaptive case.

**Comment by fdkong** [comment]
> Why does distributed run faster than replicated? I thought the only advantage distributed had over replicated was a reduction in memory.

I believe the time reduction mainly came from a distributed mesh generator. The distributed mesh generator is way faster than the serial generator since it does not need to move data around. 

  ```
[gmg]
    type = DistributedRectilinearMeshGenerator
    dim = 3
    nx = 64
    ny = 64
    nz = 64
    xmin = -20
    xmax = 20
    ymin = -20
    ymax = 20
    zmin = -20
    zmax = 20
  []
```

**Comment by fdkong** [comment, marked as answer]
It would be helpful if someone can share a graph from `gperftools` for the distributed mesh with the MA.  https://mooseframework.inl.gov/application_development/profiling.html

From @simopier's input and description, I feel like this might be related to "point ghosting functor" on which we have trouble when doing uniform refinement.  However, a confirmation is still needed to avoid doing too much unrelated optimization.

**Reply by simopier** [reply]
I would be happy to do that, but these simulations are quite expensive to run and require HPC. Unfortunately, I have not been able to install/use `gperftools` on HPC. 

Any recommendations/insight for that?

Thank you!

**Reply by fdkong** [reply]
If you are on sawtooth, you could do the following 

```
module load use.moose
module  load gperftools/2.9-gcc-8.5.0 
```

**Reply by simopier** [reply]
I performed the exact same simulations on Sawtooth with `gperftools`, but I am getting very different results then on Falcon, as you can see in the table below:

| Without mesh adaptivity      | With mesh adaptivity (larger domain)     |
|-------------|-------------|
| <img width="550" alt="Screen Shot 2021-07-06 at 8 19 55 AM" src="https://user-images.githubusercontent.com/17445651/124616281-2f1c1500-de33-11eb-8383-b6c589520e69.png"> | <img width="549" alt="Screen Shot 2021-07-06 at 8 20 07 AM" src="https://user-images.githubusercontent.com/17445651/124616320-36dbb980-de33-11eb-861f-dec04d891c68.png"> |

Now, the relative performances of the simulations are the opposite of what I was getting on falcon. Distributed mesh slows down the simulations when I am not using mesh adaptivity, but speed things up when I do use mesh adaptivity. I attach the `.prof` files of the runs on Sawtooth. I am quite confused by these results, as they seem to depend on the machine on which I perform the simulations. 

Let me know if you need additional information, or if you have any insight. 

Thank you.
[https://inlbox.box.com/s/v4a8vp2iplre9255nhnpa49ef1u82sql](https://inlbox.box.com/s/v4a8vp2iplre9255nhnpa49ef1u82sql)

**Reply by GiudGiud** [reply]
How many MPI proceses are you running on each? did you bind the processes? (--bind-to option, not usually required, but can explain some performance differences if processes migrate from one socket to the next)

**Reply by simopier** [reply]
For both Falcon and Sawtooth, I used `#PBS -l select=2:ncpus=36:mpiprocs=36`. 

I have never used the `--bind-to` option. I saw [here](https://www.open-mpi.org/faq/?category=tuning#using-paffinity-v1.4) that it can be used as `--bind-to-none` (Default), `--bind-to-core`, or as `--bind-to-socket`. Which option would you advise me to use?

**Reply by GiudGiud** [reply]
Why select=2? If in one run all the processes ended on one node and one other it was split, then that would make the communication costs very different. To know where the processes ended up, you can call `hostname` inside mpirun

I used to bind to socket with OpenMOC, to avoid inter-socket migration. Binding to cores is fine too, as long as only one process gets bound to each core.

**Reply by fdkong** [reply]
> Why select=2?

@simopier is trying to take two nodes. The nodes are not shared by anyone because he selected all 36 cores. That being said, the PBS script should not be the issue.

We pushed too far here. Let us back a bit.

Questions for @simopier 

1) Could you generate a picture of the call graph with .prof files? Nobody except you can create that picture. The .prof files depend on your executable and libs.  Again, instructions are here https://mooseframework.inl.gov/application_development/profiling.html

2) In the second picture (with mesh adaptivity), there are three lines. Were all three simulations generated using the same environment? Especially whether or not the gperf module was loaded for all three simulations. If the module were loaded only for one simulation, the simulation results would be very different. TMalloc shipped with the gperf can significantly speed up a simulation for specific problems. 

It might be worthwhile to unload the gperf module and regenerate the second picture. We are trying to figure out whether or not the gperf module changed the whole story.

**Reply by simopier** [reply]
> Could you generate a picture of the call graph with .prof files? Nobody except you can create that picture. The .prof files depend on your executable and libs. Again, instructions are here https://mooseframework.inl.gov/application_development/profiling.html

I am not sure what I am doing wrong, but I cannot manage to create that picture. 
In my folder on Sawtooth, I have a list of files named `run12_*.prof` with `*` going from 0 to `num_processor - 1`. 

I loaded `module load use.moose gperftools`, and then I have tried running:
```
pprof run12.prof
(pprof) png > run12.png
```
but I get the following error: 
```
Using remote profile at run12.prof.
Use of uninitialized value $line in substitution (s///) at /apps/moose/stack/gperftools-2.9-gcc-8.5.0/bin/pprof line 3326.
http://run12.prof/pprof/symbol doesn't exist
/var/spool/pbs/mom_priv/jobs/5202194.sawtoothpbs.SC: line 29: syntax error near unexpected token `png'
/var/spool/pbs/mom_priv/jobs/5202194.sawtoothpbs.SC: line 29: `(pprof) png > run12.png'
```

When I try:
```
pprof run12_1.prof
(pprof) png > run12_1.png
```
I get:
```
Using local file run12_1.prof.
Did not specify profile file
``` 
followed by the help message.

What am I missing?

____________

> In the second picture (with mesh adaptivity), there are three lines. Were all three simulations generated using the same environment? Especially whether or not the gperf module was loaded for all three simulations. If the module were loaded only for one simulation, the simulation results would be very different. TMalloc shipped with the gperf can significantly speed up a simulation for specific problems.
> It might be worthwhile to unload the gperf module and regenerate the second picture. We are trying to figure out whether or not the gperf module changed the whole story.

I used gperf for all 6 simulations on sawtooth. Once I figure out how to generate the gperf pictures, I will perform these simulations on sawtooth with and without gperf to check if the gperf module changed the whole story

**Reply by GiudGiud** [reply]
pprof might want the executable as an argument as well. 
```pprof --help
Usage:
/apps/moose/stack/gperftools-2.9-gcc-8.5.0/bin/pprof [options] <program> <profiles>
   <profiles> is a space separated list of profile names.
```

Please let us know how this goes

**Reply by GiudGiud** [reply]
any news on this?

**Reply by fdkong** [reply]
Sorry, I missed this.

@simopier You are almost there

> When I try:
>
```
pprof run12_1.prof
(pprof) png > run12_1.png

```
> I get:
```

Using local file run12_1.prof.
Did not specify profile file
```


Here was a small mistake. You should do `png` instead of `png >  run12_1.png`. It will generate a png file with an auto-assigned file name.

**Reply by simopier** [reply]
Hi All, 

Thank for your help! I have been able to get the visuals from gperftools as a pdf using the command:
```
/apps/moose/stack/gperftools-2.9-gcc-8.5.0/bin/pprof --pdf /home/simopa/projectsSawtooth_gp/moose/modules/combined/combined-oprof run_0.prof > run_0.pdf
```

I have re-run the simulations described above on Falcon, on Sawtooth without the gperftools, and on Sawtooth with the gperftools. The performances are shown in the table below. Once again, I show on the left the performance without Mesh adaptivity, and with adaptivity on the right. 

| Machine | Without mesh adaptivity      | With mesh adaptivity (larger domain)     |
|-------------|-------------|-------------|
| Falcon | <img width="549" alt="0_falcon_NMA" src="https://user-images.githubusercontent.com/17445651/128246958-ca7da90b-b19f-41b4-b457-f4fa5b5a0198.png"> | <img width="549" alt="0_falcon_MA" src="https://user-images.githubusercontent.com/17445651/128247210-b922952a-71d0-4611-a568-cacc387beec1.png"> |
| Sawtooth without gperftools | <img width="549" alt="0_sawtooth_NMA" src="https://user-images.githubusercontent.com/17445651/128258880-712d8c64-6cc4-44ae-b15e-b46229cb675a.png"> | <img width="547" alt="0_sawtooth_MA" src="https://user-images.githubusercontent.com/17445651/128258910-d69dde5f-7fe3-4acf-b2c0-488354b5ddb9.png"> |
| Sawtooth with gperftools | <img width="550" alt="0_sawtooth_NMA_gperf" src="https://user-images.githubusercontent.com/17445651/128258933-776b0e92-7514-4d38-ac6f-dfd1c62b9f2b.png"> | <img width="549" alt="0_sawtooth_MA_gperf" src="https://user-images.githubusercontent.com/17445651/128258970-1d0a4b3d-567c-4470-a5fb-ac516778f3a8.png"> |

Even if I ran the same simulations than I presented in previous posts, the trends changed. In the original post, I showed that using both distributed mesh and mesh adaptivity was significantly more expensive than using mesh adaptivity with replicated mesh. However, it does not seem to be the same anymore (which is good!). I'm not sure what changed, though. Maybe the differences are due to the way nodes are being used and is run-specific?

Another thing to note is that using gperftool does not seem to significantly affect the performances on sawtooth.

I also attach some of the gperftools plot here:
[gperf_plots.zip](https://github.com/idaholab/moose/files/6939482/gperf_plots.zip)

**Reply by GiudGiud** [reply]
so if I summarize:
- with increasing simulation time, distributing the mesh goes from slower to faster than replicating (**why is that**)
- ptscotch distribution makes MA slower, to the point that it's the same performance as replication (could this be an input error? did you check the ptscotch runs were actually labelled as distributed?)
- MA works faster with distributed meshes than replicated (makes sense, adaptivity effort can be distributed)

**Reply by fdkong** [reply]
Thanks so much for all these data. I did not see anything obviously wrong. 

"FEProblemBase::checkExceptionAndStopSolve" takes a nonnegligible  amount of time. That indicates: 1)  the problem is not balanced (the mesh might not be repartitioned after adaptivity?)  or 2) the machine network is so bad. 

Might try a small example to verify 1). For 2), you could use "skip_exception_check=true" in executioner block to skip this check. The check involves a global communication.

**Reply by GiudGiud** [reply]
Any updates on this? is there still an issue with the updated results?

**Comment by fdkong** [comment]
I also converted  this discussion  to a moose ticket in case nontrivial work is required to resolve the issue. 

https://github.com/idaholab/moose/issues/18191
